### DISCUSSION TITLE
Code generator support

### DISCUSSION THREAD

**Discussion by luan-xiaokun** [eclipse-langium/langium#642] [General]
The document says that a code generator can be plugged to provide a more powerful toolchain. Many DSL languages, such as modeling languages, expect code generation to get executable code, or input models for model checkers.

As a newcomer to TypeScript, I find it hard to figure out how to support code generation with the langium tool.  I am more concerned about the following problems, and any discussion on this is welcome.

1. Do I need to describe the syntax of the target language in langium's grammar language, or is there another way?
2. Where should I plug the code generator, for example as an injected service, or as another extension?
3. Can langium provide a way to map the syntax rules of two languages?
4. Are there other language tools that can support code generation?

Since langium development is probably still at a relatively early stage, I have a feeling there may not be direct support, but I would love to have such features.

**Comment by msujew** [comment]
Hey @luan-xiaokun,

> Do I need to describe the syntax of the target language in langium's grammar language, or is there another way?

As of now, there's no other way to make use of Langium without defining your language in Langium's grammar language. The framework relies on the structural and syntactical information provided in your grammar to work.

> Where should I plug the code generator, for example as an injected service, or as another extension?

Sure, having your generator as an injected service sounds perfectly reasonable. Especially, since you'll have access to all your other services in there as well.

> Can langium provide a way to map the syntax rules of two languages?

You mean like running two languages in a language server and have them interact with each other? Yes, although we don't have documentation on that yet, this feature has been available for quite some time (see https://github.com/langium/langium/pull/311). You can simply have both languages in your langium config file (which is why the `languages` object is an array).

> Are there other language tools that can support code generation?

Langium is the spiritual successor of [Eclipse Xtext](https://www.eclipse.org/Xtext/) which has a more opinionated approach to code generation. We wanted to give users more options on how to they want to generate their code. For code generation in Langium itself, you can use every kind of text templating system that you like.

> Since langium development is probably still at a relatively early stage, I have a feeling there may not be direct support, but I would love to have such features.

After almost 18 months of development we believe that the framework is already in a good position. We use it in several production-ready projects already. For code generation support especially, we tried to keep it hands-off and let users implement it the way they want.

**Reply by luan-xiaokun** [reply]
Thanks for your reply! The information is really helpful to me.

> You mean like running two languages in a language server and have them interact with each other?

Please allow me to rephrase my question. Suppose we have a source language (such as a modeling language) and a target language (such as normal programming language) , and sometimes the target language does not support all features of source language. The code generation requires some rules to map the elements in the source language to the target language.

Here's what I think is one way to implement code generation using Langium. We have the AST of a program of the source language, and we want to use the rules to build an AST of the target language based on this, and then we can export a target language program text file. Please let me know if I'm wrong about certain steps.

The interaction seems to be one-way, so I was wondering if we can just write some syntax rules in the grammar language (extended) to define such mapping rules, and then hand over the implementation of code generation between source language and target language to another "generator of code generator", just like langium uses a generator for the language syntax.

This idea may be naive, as the mapping between languages can be very complex and situation dependent. But I will probably want to use this approach in my case, because we have a modeling language and may have multiple target languages to generate code.

**Reply by msujew** [reply]
Alright, I see. Thanks for giving more context on your question.

So, if I understand you correctly, you're trying to build another language which allows you to define a mapping between source and target languages, right? Depending on how complex the mapping becomes and how many target languages you aim to support, this sounds quite resonable. Xtext used to do something similar (called Xbase), in which you were able to interop with native Java code, allowing you to create a very simple DSL to Java code generator.

However, regarding your initial question: This isn't something that we plan to support in Langium. We always try to keep the framework as lean and simple as possible. But nothing stops you from creating your own DSL for building such a mapping.

**Reply by luan-xiaokun** [reply]
> you're trying to build another language which allows you to define a mapping between source and target languages, right?

Yes. I think this approach is more flexible, but also more complex. Another possible approach is creating an interface / function in typescript.

> This isn't something that we plan to support in Langium

I get it, you have explained Langium's goals very well. Thanks again!
